crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1889,49 +1891,52 @@ impl<'db> Type<'db> {
         }
     }
 
-    /// Return the type of `len()` on a type if it is known more precisely than `int`,
+    /// Return the type of calling the `__len__` method on a type.
+    fn __len__(&self, db: &'db dyn Db) -> CallDunderLenOutcome<'db> {
+        let statically_known_len = match self {
+            Type::BytesLiteral(bytes) => Some(bytes.python_len(db)),
+            Type::StringLiteral(string) => Some(string.python_len(db)),
+            Type::Tuple(tuple) => Some(tuple.len(db)),
+            _ => None,
+        };
+
+        if let Some(len) = statically_known_len {
+            return CallDunderLenOutcome::StaticallyKnown(len);
+        }
+
+        match self.call_dunder(db, "__len__", &CallArguments::positional([*self])) {
+            CallDunderOutcome::MethodNotAvailable => CallDunderLenOutcome::MethodNotAvailable,
+            CallDunderOutcome::Call(outcome) => CallDunderLenOutcome::Call(outcome),
+            CallDunderOutcome::PossiblyUnbound(outcome) => {
+                CallDunderLenOutcome::PossiblyUnbound(outcome)
+            }
```

Comment:
What's sort of confusing with `CallDunderOutcome` is that a `CallOutcome` also has a `PossiblyUnbound` variant but I don't think they're the same?